### PR TITLE
Verify services deleted when cluster scales down

### PR DESCRIPTION
**Pull request by russgold**
This addresses bug https://jira.oraclecorp.com/jira/browse/OWLS-71460, which reports that service services were not being removed on scale down. The unit tests now show that they are, which may mean that the service work over the past month has addressed the issue.

This change now addresses a number of code issues, including demonstrating the ability to unit test cluster scaling and file I/O (without actually accessing the file system). It also does some code cleanup, needed to get the unit tests to work, including:

- Moved certificate reading from the REST service to the Certificates class, so we no longer need to bring up a server to read them,
- Moved the JobWatcher cache into the JobWatcher class, so we don't need to pass parameters through a long chain of steps, most of which don't do anything with them, and
- Used a delegate to break the dependency of DomainProcessorImpl on Main, allowing it to be unit-tested.

### CODE REVIEW

File: operator/src/test/java/oracle/kubernetes/operator/JobWatcherTest.java [hunk 1]
```diff
@@ -9,6 +9,8 @@
 import static org.hamcrest.Matchers.both;
 import static org.hamcrest.Matchers.hasEntry;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
```
Comment (markxnelson) [inline operator/src/test/java/oracle/kubernetes/operator/JobWatcherTest.java]: update copyright notice

File: operator/src/test/java/oracle/kubernetes/operator/helpers/KubernetesTestSupport.java [hunk 2]
```diff
@@ -290,12 +333,17 @@ T createResource(String namespace, T resource) {
     }
 
     Object listResources(String namespace, String fieldSelector, String... labelSelectors) {
-      throw new UnsupportedOperationException("list operation not supported");
+      if (listFactory == null)
```
Comment (markxnelson) [inline operator/src/test/java/oracle/kubernetes/operator/helpers/KubernetesTestSupport.java]: braces

File: operator/src/test/java/oracle/kubernetes/operator/helpers/TuningParametersStub.java [hunk 3]
```diff
@@ -39,6 +39,11 @@ public PodTuning getPodTuning() {
         LIVENESS_PERIOD);
   }
 
+  @Override
+  public MainTuning getMainTuning() {
```
Comment (markxnelson) [inline operator/src/test/java/oracle/kubernetes/operator/helpers/TuningParametersStub.java]: what do all these values mean?

File: operator/src/test/java/oracle/kubernetes/operator/work/FiberTestSupport.java [hunk 4]
```diff
@@ -70,6 +86,10 @@ public boolean hasItemScheduledAt(int time, TimeUnit unit) {
     return schedule.containsItemAt(time, unit);
   }
 
+  public Engine getEngine() {
```
Comment (markxnelson) [inline operator/src/test/java/oracle/kubernetes/operator/work/FiberTestSupport.java]: javadoc

File: operator/src/main/java/oracle/kubernetes/operator/DomainProcessor.java [hunk 5]
```diff
@@ -14,10 +14,6 @@
 
 public interface DomainProcessor {
 
-  public static DomainProcessor getInstance() {
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/DomainProcessor.java]: whoops. Fixing now. I think we should have a pre-commit hook that checks these.

File: operator/src/main/java/oracle/kubernetes/operator/DomainProcessorDelegate.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+package oracle.kubernetes.operator;
+
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import oracle.kubernetes.operator.helpers.KubernetesVersion;
+import oracle.kubernetes.operator.work.FiberGate;
+import oracle.kubernetes.operator.work.Step;
+
+public interface DomainProcessorDelegate {
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/DomainProcessorDelegate.java]: Added

File: operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java [hunk 7]
```diff
@@ -148,67 +146,64 @@ public void dispatchPodWatch(Watch.Response<V1Pod> item) {
...
-                  // Pod was deleted, but sko still contained a non-null entry
-                  LOGGER.info(
-                      MessageKeys.POD_DELETED, domainUID, metadata.getNamespace(), serverName);
-                  makeRightDomainPresence(existing, true, false, true);
-                }
-                break;
+                            }
+                            return current;
+                          });
+              if (oldPod != null && !existing.isDeleting()) {
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java]: Good catch.

File: operator/src/main/java/oracle/kubernetes/operator/JobWatcher.java [hunk 8]
```diff
@@ -10,12 +10,14 @@
 import io.kubernetes.client.models.V1JobStatus;
 import io.kubernetes.client.models.V1ObjectMeta;
 import io.kubernetes.client.util.Watch;
+import java.util.HashMap;
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/JobWatcher.java]: Done

File: operator/src/main/java/oracle/kubernetes/operator/JobWatcher.java [hunk 9]
```diff
@@ -28,19 +30,36 @@
...
-  private final String ns;
+  private final String namespace;
 
   // Map of Pod name to Complete
   private final ConcurrentMap<String, Complete> completeCallbackRegistrations =
       new ConcurrentHashMap<>();
 
   /**
-   * Factory for JobWatcher.
+   * If a JobWatcher has already been created, returns it.
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/JobWatcher.java]: Oops. That was left over from a previous implementation. Now fixed to indicate that it is never null.

File: operator/src/main/java/oracle/kubernetes/operator/helpers/JobHelper.java [hunk 10]
```diff
@@ -115,32 +111,18 @@ Domain getDomain() {
   /**
    * Factory for {@link Step} that creates WebLogic domain introspector job.
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/JobHelper.java]: I don't see it. What param is missing?

File: operator/src/main/java/oracle/kubernetes/operator/helpers/JobHelper.java [hunk 11]
```diff
@@ -167,10 +149,7 @@ private static boolean runIntrospector(Packet packet, DomainPresenceInfo info) {
     LOGGER.fine("runIntrospector topology: " + config);
     LOGGER.fine("runningServersCount: " + runningServersCount(info));
     LOGGER.fine("creatingServers: " + creatingServers(info));
-    if (config == null || (runningServersCount(info) == 0 && creatingServers(info))) {
-      return true;
-    }
-    return false;
+    return config == null || (runningServersCount(info) == 0 && creatingServers(info));
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/JobHelper.java]: I agree that it is insufficiently clear, and have tried to improve it.

File: operator/src/main/java/oracle/kubernetes/operator/logging/LoggingFactory.java [hunk 12]
```diff
@@ -15,7 +15,7 @@
   // map from resourceBundleName to facade
   private static final Map<String, LoggingFacade> facade = new HashMap<String, LoggingFacade>();
 
-  private static JSON json = null;
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/logging/LoggingFactory.java]: Fixed

File: operator/src/test/java/oracle/kubernetes/operator/JobWatcherTest.java (hunk 1 above)
Comment (russgold) [reply operator/src/test/java/oracle/kubernetes/operator/JobWatcherTest.java]: Done

File: operator/src/main/java/oracle/kubernetes/operator/helpers/JobHelper.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (markxnelson) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/JobHelper.java]: might be just an artifact of the diff... looks like a few were deleted.

[the remark this answers is not in the document]
Comment (markxnelson) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/JobHelper.java]: ah no, i see now - my bad - i read the wrong method signature

File: operator/src/test/java/oracle/kubernetes/operator/helpers/KubernetesTestSupport.java (hunk 2 above)
Comment (russgold) [reply operator/src/test/java/oracle/kubernetes/operator/helpers/KubernetesTestSupport.java]: I disagree. I thought you'd read the book. Did you not get to Uncle Bob's explanation for why it is better not to use them in such cases?

File: operator/src/test/java/oracle/kubernetes/operator/helpers/TuningParametersStub.java (hunk 3 above)
Comment (russgold) [reply operator/src/test/java/oracle/kubernetes/operator/helpers/TuningParametersStub.java]: It's a test stub. I had to use values to get it to compile. I think the original code is needlessly confusing, but have had no real reason to fix it.

File: operator/src/test/java/oracle/kubernetes/operator/helpers/KubernetesTestSupport.java (hunk 2 above)
Comment (markxnelson) [reply operator/src/test/java/oracle/kubernetes/operator/helpers/KubernetesTestSupport.java]: i read it cover to cover and i tried really hard to find that passage, but i was not successful. i would love to see his argument though if you can point me to it

File: operator/src/test/java/oracle/kubernetes/operator/work/FiberTestSupport.java (hunk 4 above)
Comment (russgold) [reply operator/src/test/java/oracle/kubernetes/operator/work/FiberTestSupport.java]: Done

**Comment by russgold** [comment]
At least one person has thought that adding a label:
```
weblogic.createdByOperator: “false”
```
should be enough to tell the operator to leave a resource alone. It seems a reasonable assumption, so I have updated the code to behave accordingly.

**Pull request merged**